dwave/optimization/generators.py

```diff
@@ -17,120 +17,64 @@
-            raise ValueError(f"`{argname}` must be a 1D array-like of non-negative numbers")
-        
-        if not np.issubdtype(array.dtype, np.number):
-            raise ValueError(f"`{argname}` must be a 1D array-like of non-negative numbers")
+    try:
+        array = np.asarray(array_like, order="C", dtype=dtype)
+    except (ValueError, TypeError) as err:
+        raise ValueError(f"`{argname}` must be an array-like of numbers") from err
 
-        if array.ndim != 1 or (array < 0).any():
-            raise ValueError(f"`{argname}` must be a 1D array-like of non-negative numbers")
+    if not np.issubdtype(array.dtype, np.number) or np.issubdtype(array.dtype, np.complexfloating):
+        raise ValueError(f"`{argname}` must be an array-like of numbers")
 
-        yield array
+    # Disallow inf and NaN. We could make this toggleable with a kwarg if there
+    # is eventually a reason for it
+    try:
+        array = np.asarray_chkfinite(array)
+    except ValueError as err:
+        raise ValueError(f"'{argname}' must not contain infs or NaNs") from err
 
+    if ndim is not None and array.ndim != ndim:
+        # we don't "size down"
+        if array.ndim > ndim:
+            raise ValueError(f"'{argname}' must be a {ndim}D array-like")
 
-def _2d_nonnegative_int_array(**kwargs: numpy.typing.ArrayLike) -> typing.Iterator[np.ndarray]:
-    """Coerce all given array-likes to 2d NumPy arrays of non-negative integers and
-    raise a consistent error message if it cannot be cast.
-
-    Keyword argument names must match the argument names of the calling function
-    for the error message to make sense.
-    """
-    for argname, array in kwargs.items():
-        try:
-            array = np.atleast_2d(np.asarray(array, dtype=int))
-        except (ValueError, TypeError):
-            raise ValueError(f"`{argname}` must be a 2d array-like of non-negative integers")
-        
-        if not np.issubdtype(array.dtype, np.number):
-            raise ValueError(f"`{argname}` must be a 2d array-like of non-negative integers")
-
-        if array.ndim != 2 or (array < 0).any():
-            raise ValueError(f"`{argname}` must be a 2d array-like of non-negative integers")
-
-        yield array
-
-
-def _2d_nonnegative_square_array(**kwargs: numpy.typing.ArrayLike) -> typing.Iterator[np.ndarray]:
-    """Coerce all given array-likes to 2D NumPy arrays of non-negative floats
-    or integers and raise a consistent error message if it cannot be cast.
-
-    Keyword argument names must match the argument names of the calling function
-    for the error message to make sense.
-    """
-    for argname, array in kwargs.items():
-        try:
-            array = np.atleast_2d(np.asarray(array))
-        except (ValueError, TypeError):
-            raise ValueError(f"`{argname}` must be a square 2D array-like of non-negative numbers")
+        # we do "size up"
+        while array.ndim < ndim:
+            array = array[np.newaxis]
```

Comment:
I'm not quite sure what is intended with this clever trick but I doubt it is this:

```
model = job_shop_scheduling(times=[1, 2, 3, 4], machines=[4, 3, 2, 1])
ValueError: arrays must have the same shape
```

I think the code will be clearer if, for a requirement on square matrices, you check that `array.shape[0] == array.shape[1]` and don't upsize.

Reply:
That error was present in the original code. Turns out it fails on `main` for every non-square times/machines. Will fix.

Re the dimension increase, I am following the pattern of [np.atleast_2d()](https://numpy.org/doc/stable/reference/generated/numpy.atleast_2d.html), which is typical in this sort of place.

Reply:
Ah, that's a different bug. Will fix that one too.

Reply:
https://github.com/dwavesystems/dwave-optimization/issues/43